### PR TITLE
add ability to set system parameter defaults dynamically

### PR DESCRIPTION
**Pull request by guswynn**
See https://materializeinc.slack.com/archives/C0246GEHL8N/p1683835189645309 for more information. 

Bootstrapped system vars are overridden by resets. This makes them hard to use with testdrive. This pr fixes that, by making dynamic defaults explicitly supported, and renames things to make it more clear

### Motivation

  * This PR fixes a recognized bug.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

### CODE REVIEW

File: src/adapter/src/coord/ddl.rs [hunk 1]
```diff
@@ -192,6 +192,14 @@ impl Coordinator {
...
                     update_metrics_retention |= name == vars::METRICS_RETENTION.name();
                 }
+                catalog::Op::ResetAllSystemConfiguration => {
+                    // Assume they all need to be updated.
+                    // We could see if the config's have actually changed, but
+                    // this is simpler.
+                    update_compute_config = true;
+                    update_storage_config = true;
+                    update_metrics_retention = true;
+                }
```
Comment (guswynn) [inline src/adapter/src/coord/ddl.rs:202]: This is another bug, where `RESET ALL` wasn't propagated correctly.

File: misc/python/materialize/mzcompose/services.py [hunk 2]
```diff
@@ -65,7 +65,8 @@ def __init__(
         restart: Optional[str] = None,
         use_default_volumes: bool = True,
         ports: Optional[List[str]] = None,
-        bootstrap_system_parameters: Optional[List[str]] = None,
+        cli_system_variable_defaults: Optional[List[str]] = None,
+        additional_cli_system_variable_defaults: Optional[List[str]] = None,
```
Comment (benesch) [inline misc/python/materialize/mzcompose/services.py:69]: `extra` is our convention for mzcompose service parameters like this.

```suggestion
        extra_cli_system_variable_defaults: Optional[List[str]] = None,
```

File: misc/python/materialize/checks/mzcompose_actions.py [hunk 3]
```diff
@@ -30,11 +30,11 @@ def __init__(
         self,
         tag: Optional[MzVersion] = None,
         environment_extra: List[str] = [],
-        bootstrap_system_parameters: Optional[List[str]] = None,
+        cli_system_variable_defaults: Optional[List[str]] = None,
```
Comment (benesch) [inline misc/python/materialize/checks/mzcompose_actions.py:33]:
```suggestion
        system_parameter_defaults: Optional[List[str]] = None,
```

File: misc/python/materialize/checks/mzcompose_actions.py [hunk 4]
```diff
@@ -30,11 +30,11 @@ def __init__(
...
     ) -> None:
         self.tag = tag
         self.environment_extra = environment_extra
-        self.bootstrap_system_parameters = bootstrap_system_parameters
+        self.cli_system_variable_defaults = cli_system_variable_defaults
```
Comment (benesch) [inline misc/python/materialize/checks/mzcompose_actions.py:37]:
```suggestion
        self.system_parameter_defaults = cli_system_parameter_defaults
```

File: misc/python/materialize/checks/mzcompose_actions.py [hunk 5]
```diff
@@ -45,7 +45,7 @@ def execute(self, e: Executor) -> None:
             image=image,
             external_cockroach=True,
             environment_extra=self.environment_extra,
-            bootstrap_system_parameters=self.bootstrap_system_parameters,
+            cli_system_variable_defaults=self.cli_system_variable_defaults,
```
Comment (benesch) [inline misc/python/materialize/checks/mzcompose_actions.py:48]:
```suggestion
            cli_system_parameter_defaults=self.cli_system_parameter_defaults,
```

File: misc/python/materialize/mzcompose/services.py (hunk 2 above)
Comment (benesch) [inline misc/python/materialize/mzcompose/services.py:69]:
```suggestion
        system_parameter_defaults: Optional[List[str]] = None,
        additional_system_parameter_defaults: Optional[List[str]] = None,
```

File: misc/python/materialize/mzcompose/services.py [hunk 6]
```diff
@@ -93,16 +94,19 @@ def __init__(
             *environment_extra,
         ]
 
-        if bootstrap_system_parameters is None:
-            bootstrap_system_parameters = [
+        if cli_system_variable_defaults is None:
+            cli_system_variable_defaults = [
```
Comment (benesch) [inline misc/python/materialize/mzcompose/services.py:98]:
```suggestion
        if system_parameter_defaults is None:
            system_parameter_defaults = [
```

File: misc/python/materialize/mzcompose/services.py [hunk 7]
```diff
@@ -93,16 +94,19 @@ def __init__(
...
                 "persist_sink_minimum_batch_updates=128",
                 "enable_multi_worker_storage_persist_sink=true",
                 "storage_persist_sink_minimum_batch_updates=100",
             ]
 
-        if len(bootstrap_system_parameters) > 0:
+        if additional_cli_system_variable_defaults is not None:
+            cli_system_variable_defaults += additional_cli_system_variable_defaults
```
Comment (benesch) [inline misc/python/materialize/mzcompose/services.py:105]:
```suggestion
        if extra_system_parameter_defaults is not None:
            system_parameter_defaults += extra_system_parameter_defaults
```

File: misc/python/materialize/mzcompose/services.py [hunk 8]
```diff
@@ -93,16 +94,19 @@ def __init__(
...
+
+        if len(cli_system_variable_defaults) > 0:
```
Comment (benesch) [inline misc/python/materialize/mzcompose/services.py:107]:
```suggestion
        if len(system_parameter_defaults) > 0:
```

File: misc/python/materialize/mzcompose/services.py [hunk 9]
```diff
@@ -93,16 +94,19 @@ def __init__(
...
             environment += [
-                "MZ_BOOTSTRAP_SYSTEM_PARAMETER=" + ";".join(bootstrap_system_parameters)
+                "MZ_SYSTEM_VAR_DEFAULT=" + ";".join(cli_system_variable_defaults)
```
Comment (benesch) [inline misc/python/materialize/mzcompose/services.py:109]:
```suggestion
                "MZ_SYSTEM_PARAMETER_DEFAULT=" + ";".join(system_parameter_defaults)
```

File: src/environmentd/src/bin/environmentd/main.rs [hunk 10]
```diff
@@ -446,15 +446,15 @@ pub struct Args {
...
     )]
     bootstrap_builtin_cluster_replica_size: String,
-    /// An list of NAME=VALUE pairs for bootstrapping system parameters that are
-    /// not already modified.
+    /// An list of NAME=VALUE pairs used to override static defaults
+    /// for system parameters.
     #[clap(
         long,
-        env = "BOOTSTRAP_SYSTEM_PARAMETER",
+        env = "SYSTEM_VAR_DEFAULT",
```
Comment (benesch) [inline src/environmentd/src/bin/environmentd/main.rs:453]:
```suggestion
        env = "SYSTEM_PARAMETER_DEFAULT",
```

File: src/environmentd/src/bin/environmentd/main.rs [hunk 11]
```diff
@@ -446,15 +446,15 @@ pub struct Args {
...
         multiple = true,
         value_delimiter = ';'
     )]
-    bootstrap_system_parameter: Vec<KeyValueArg<String, String>>,
+    system_var_default: Vec<KeyValueArg<String, String>>,
```
Comment (benesch) [inline src/environmentd/src/bin/environmentd/main.rs:457]:
```suggestion
    system_parameter_default: Vec<KeyValueArg<String, String>>,
```

File: src/environmentd/src/bin/environmentd/main.rs [hunk 12]
```diff
@@ -796,8 +796,8 @@ fn run(mut args: Args) -> Result<(), anyhow::Error> {
         default_storage_cluster_size: args.default_storage_host_size,
         bootstrap_default_cluster_replica_size: args.bootstrap_default_cluster_replica_size,
         bootstrap_builtin_cluster_replica_size: args.bootstrap_builtin_cluster_replica_size,
-        bootstrap_system_parameters: args
-            .bootstrap_system_parameter
+        cli_system_variable_defaults: args
+            .system_var_default
```
Comment (benesch) [inline src/environmentd/src/bin/environmentd/main.rs:800]:
```suggestion
            .system_parameter_default
```

File: test/testdrive/mzcompose.py [hunk 13]
```diff
@@ -82,7 +82,13 @@ def workflow_default(c: Composition, parser: WorkflowArgumentParser) -> None:
         validate_postgres_stash="materialized",
     )
 
-    materialized = Materialized(default_size=args.default_size)
+    materialized = Materialized(
+        default_size=args.default_size,
+        # Used to test some behavior in `system-param-defaults.td`.
+        additional_cli_system_variable_defaults=[
```
Comment (benesch) [inline test/testdrive/mzcompose.py:88]:
```suggestion
        extra_system_parameter_defaults=[
```

File: test/upgrade-matrix/nodes.py [hunk 14]
```diff
@@ -41,7 +41,7 @@ def actions(self, scenario: Scenario) -> List[Action]:
         # As this action may need start very old Mz versions,
         # we do not use any bootstrap_systme_parameters
         return [
-            StartMz(tag=self.version, bootstrap_system_parameters=[]),
+            StartMz(tag=self.version, cli_system_variable_defaults=[]),
```
Comment (benesch) [inline test/upgrade-matrix/nodes.py:44]:
```suggestion
            StartMz(tag=self.version, system_parameter_defaults=[]),
```

File: test/upsert/mzcompose.py [hunk 15]
```diff
@@ -68,7 +68,7 @@ def workflow_default(c: Composition, parser: WorkflowArgumentParser) -> None:
         default_size=args.default_size,
         options=[
             "--orchestrator-process-scratch-directory=/mzdata/source_data"
-            "--bootstrap-system-parameter=upsert_source_disk_default=true"
+            "--system-var-default=upsert_source_disk_default=true"
```
Comment (benesch) [inline test/upsert/mzcompose.py:71]:
```suggestion
            "--system-parameter-default=upsert_source_disk_default=true"
```

**Review by benesch — APPROVED** [review]
Nice! So sorry but I forgot that we try to be consistent in referring to these as "system parameters" in our public API (so our command line flags), since that's the PostgreSQL terms. The internal code is super inconsistent about param vs var vs variable but that's definitely a fight for another day.

Thanks very much for shaving this yak. This new behavior is _so_ much better than the old behavior of `--bootstrap-system-var`. PR turned out really clean, too!

**Pull request merged**